### PR TITLE
add qwenvl

### PR DESCRIPTION
**Pull request by zouxinyi0625**

### CODE REVIEW

File: chatlearn/algorithm/grpo_utils/policy_trainer.py [hunk 1]
```diff
@@ -99,16 +104,24 @@ def preprocess_data_list(self, data_list: List[Dict[str, Any]], training: bool):
             tokens_ = data_b["all_tokens"].long()
             prompt_token_length = data_b["prompt_token_length"]
             response_token_length = data_b["response_token_length"]
+            position_ids = data_b.get("position_ids", None)
```
Comment (haiasd) [inline chatlearn/algorithm/grpo_utils/policy_trainer.py:107]: 加个注释说明下下面是给vl的

File: chatlearn/algorithm/grpo_utils/policy_trainer.py [hunk 2]
```diff
@@ -158,16 +182,30 @@ def train_step(self, data_list: List[Dict[str, Any]], **kwargs): # pylint: disab
...
-                attention_mask=None,
-                position_ids=inputs["position_ids"],
-                use_cache=False,
-            )
+            if "pixel_values" in inputs:
+                output = self.model(
+                    input_ids=inputs['all_tokens'],
+                    pixel_values=inputs['pixel_values'],
+                    image_grid_thw=inputs['image_grid_thw'],
+                    attention_mask=None,#inputs['attention_mask'],
```
Comment (haiasd) [inline chatlearn/algorithm/grpo_utils/policy_trainer.py:196]: 后面注释可以删掉了

File: chatlearn/algorithm/grpo_utils/policy_trainer.py [hunk 3]
```diff
@@ -158,16 +182,30 @@ def train_step(self, data_list: List[Dict[str, Any]], **kwargs): # pylint: disab
...
+                    position_ids=inputs['position_ids'],
+                    use_cache=False,
+                    rope_deltas=None
+                )
+            else:
+                output = self.model(
+                    input_ids=inputs['all_tokens'],
+                    attention_mask=None,#inputs['attention_mask'],
```
Comment (haiasd) [inline chatlearn/algorithm/grpo_utils/policy_trainer.py:204]: 同上

File: chatlearn/algorithm/grpo_utils/trainer_utils.py [hunk 4]
```diff
@@ -54,7 +57,12 @@ def sp_split(input_tensor:torch.Tensor, split_dim:int, sp_size:int, sp_local_ran
     """
     return torch.chunk(input_tensor, sp_size, split_dim)[sp_local_rank]
 
-def generate_loss_mask_position_ids(tokens: torch.Tensor, prompt_token_length: List[int], response_token_length: List[int]):
+def generate_loss_mask_position_ids(
+    tokens: torch.Tensor,
+    prompt_token_length: List[int],
+    response_token_length: List[int],
+    prompt_position_ids: List[List[int]]
```
Comment (haiasd) [inline chatlearn/algorithm/grpo_utils/trainer_utils.py:64]: 加个默认值 None

File: chatlearn/data/vl_prompt_dataset.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+from typing import List, Dict
+import re
+from torch.utils.data import Dataset
+from transformers import AutoTokenizer, AutoProcessor
+from chatlearn.data.vision_utils import process_image, process_video
+from chatlearn.models.patches.transformers.qwen2_5_vl_patch import get_rope_index
+
+
+class PromptPipeline(Dataset):
+    """
```
Comment (haiasd) [inline chatlearn/data/vl_prompt_dataset.py:12]: 改成vl的示例

File: scripts/base_env.sh [hunk 6]
```diff
@@ -4,8 +4,7 @@ ray stop
 
 export CUDA_DEVICE_MAX_CONNECTIONS=1
 export NCCL_DEBUG=WARN
-
-
+# export GPUS_PER_NODE=2
```
Comment (haiasd) [inline scripts/base_env.sh:7]: 注释可以删掉

File: docs/zh/tutorial/tutorial_grpo_fsdp_qwenvl.md [hunk 7]
```diff
@@ -0,0 +1,76 @@
+# 基于 FSDP 的端到端 Qwen2.5VL GRPO训练流程
+
+本文档提供使用 ChatLearn、PyTorch FSDP 和 vLLM 框架来对Qwen2.5-VL模型进行GRPO训练的快速开始指南。
+
+## 环境配置
+1. Docker镜像准备
+我们建议在PAI [DSW](https://help.aliyun.com/zh/pai/user-guide/create-and-manage-dsw-instances/)/[DLC](https://help.aliyun.com/zh/pai/user-guide/create-a-training-task?spm=a2c4g.11186623.help-menu-30347.d_3_3_5_5.2dfb1925l3QjwG)中运行该示例，你需要填写如下镜像地址来启动实例：
+```bash
+dsw-registry.cn-shanghai.cr.aliyuncs.com/pai-training-algorithm/chatlearn:torch2.6.0-vllm0.8.5-ubuntu24.04-cuda12.6-py312-v1.0
```
Comment (haiasd) [inline docs/zh/tutorial/tutorial_grpo_fsdp_qwenvl.md:9]: 确认没问题可以直接覆盖原镜像

File: chatlearn/algorithm/grpo_utils/policy_trainer.py [hunk 8]
```diff
@@ -35,7 +37,10 @@
                             generate_loss_mask_position_ids,
                             split_microbatch,
                             batching,
-                            split_and_unpadding)
+                            split_and_unpadding,
+                            unpad_input)
+bert_padding.unpad_input = unpad_input
```
Comment (yytang0204) [inline chatlearn/algorithm/grpo_utils/policy_trainer.py:42]: try comment

File: chatlearn/data/vl_prompt_dataset.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+                "raw_input_ids": raw_input_ids,
+                "prompt": raw_prompt,
+                "position_ids": position_ids.squeeze().tolist(),
+                "rope_deltas": rope_deltas,
+                "data_source": data_source,
+                "ground_truth": ground_truth,
+                "multi_modal_data": multi_modal_data,
+                "mm_processor_kwargs": mm_processor_kwargs,
+                "pixel_values": pixel_values,
+                "image_grid_thw": image_grid_thw
```
[the remark this answers is not in the document]
Comment (zouxinyi0625) [reply chatlearn/data/vl_prompt_dataset.py:128]: 放在 batching里了 和 pixel_values 一致

File: docs/en/index.rst [hunk 10]
```diff
@@ -21,6 +21,7 @@ ChatLearn Documentation
 
    tutorial/tutorial_grpo_fsdp
    tutorial/tutorial_grpo_mcore
+   tutorial/tutorial_grpo_fsdp_qwenvl
```
[the remark this answers is not in the document]
Comment (zouxinyi0625) [reply docs/en/index.rst:24]: done

File: chatlearn/algorithm/grpo_utils/policy_trainer.py (hunk 8 above)
Comment (zouxinyi0625) [reply chatlearn/algorithm/grpo_utils/policy_trainer.py]: done

File: chatlearn/models/patches/transformers/qwen2_5_vl_patch.py [hunk 11]
```diff
@@ -0,0 +1,385 @@
...
+    cu_seq_lens_k: Optional[torch.LongTensor] = None,
+    max_length_q: Optional[int] = None,
+    max_length_k: Optional[int] = None,
+    target_dtype: Optional[torch.dtype] = None,
+    **kwargs, # pylint: disable=unused-argument
+):
+    """
+    Calls the forward method of Flash Attention - if the input hidden states contain at least one padding token
+    first unpad the input, then computes the attention scores and pad the final attention scores.
+    """
```
[the remark this answers is not in the document]
Comment (zouxinyi0625) [reply chatlearn/models/patches/transformers/qwen2_5_vl_patch.py:174]: done

File: chatlearn/models/patches/transformers/qwen2_5_vl_patch.py [hunk 12]
```diff
@@ -0,0 +1,385 @@
...
+        processor,
+        input_ids: Optional[torch.LongTensor] = None,
+        image_grid_thw: Optional[torch.LongTensor] = None,
+        video_grid_thw: Optional[torch.LongTensor] = None,
+        second_per_grid_ts: Optional[torch.Tensor] = None,
+        attention_mask: Optional[torch.Tensor] = None,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+    """
+    Use processor to compute the position ids for RoPE
+    """
```
[the remark this answers is not in the document]
Comment (zouxinyi0625) [reply chatlearn/models/patches/transformers/qwen2_5_vl_patch.py:28]: done

File: chatlearn/algorithm/grpo_utils/policy_trainer.py (hunk 2 above)
Comment (zouxinyi0625) [reply chatlearn/algorithm/grpo_utils/policy_trainer.py]: done

File: chatlearn/algorithm/grpo_utils/policy_trainer.py [hunk 13]
```diff
@@ -99,16 +104,24 @@ def preprocess_data_list(self, data_list: List[Dict[str, Any]], training: bool):
...
-
                 position_ids, *_ = unpad_input(position_ids.unsqueeze(-1).cuda(), attn_mask.cuda())
-                position_ids = position_ids.permute(1, 0).cpu() # For compatible with transformers
+                if len(position_ids.shape)==3:
+                    # vl
+                    position_ids = position_ids.permute(0, 2, 1).cpu()
+                else:
+                    position_ids = position_ids.permute(1, 0).cpu() # For compatible with transformers
 
             if self.sp_size > 1:
```
[the remark this answers is not in the document]
Comment (zouxinyi0625) [reply chatlearn/algorithm/grpo_utils/policy_trainer.py:124]: done in create_model

File: chatlearn/algorithm/grpo_utils/trainer_utils.py (hunk 4 above)
Comment (zouxinyi0625) [reply chatlearn/algorithm/grpo_utils/trainer_utils.py]: done

File: chatlearn/algorithm/grpo_utils/trainer_utils.py [hunk 14]
```diff
@@ -125,3 +150,43 @@ def split_and_unpadding(input_tensor: torch.Tensor, attention_mask: torch.Tensor
         input_tensor[i, :valid_seq[i]] for i in range(input_tensor.shape[0])
     ]
     return tensor_list
+
+
+
```
[the remark this answers is not in the document]
Comment (zouxinyi0625) [reply chatlearn/algorithm/grpo_utils/trainer_utils.py]: done

File: chatlearn/models/patches/transformers/qwen2_5_vl_patch.py [hunk 15]
```diff
@@ -0,0 +1,385 @@
...
+    indices_q = torch.arange(position_ids.size(0), device=position_ids.device, dtype=torch.int32)
+
+    cu_seq_lens = torch.cat(
+        (
+            indices_q[position_ids == 0],
+            torch.tensor(position_ids.size(), device=position_ids.device, dtype=torch.int32),
+        )
+    )
+
+    # =========================================================================
```
[the remark this answers is not in the document]
Comment (zouxinyi0625) [reply chatlearn/models/patches/transformers/qwen2_5_vl_patch.py:281]: done

**Review by haiasd — DISMISSED** [review]

**Review by yytang0204 — APPROVED** [review]

**Review by haiasd — APPROVED** [review]
